### ISSUE TITLE
How to embed an svg in a template from a local file (`include` and `load_data` don't seem to work for this)

### ISSUE THREAD

**Issue by rib** [Keats/gutenberg#2029]
# Documentation issue

## Summary
Hi, I'm trying to switch my blog over to using Zola, and something I was doing previously was directly embedding a few svgs (for social media icons) which I can't figure out how to do currently with Tera's `include` directive or via `set data = load_data()`

When I try to something like `{% include "github.svg" %}` (and trying lots of alternative locations for the svg) it seems to always report an error like:
```
Error: Reason: Failed to render 'post-page.html'
Error: Reason: Template '[github.svg]' not found
```

If I instead try something like:
```xml
          {% set data = load_data(path="github.svg", format="xml") %}
          <div class="icon">{{ data }}</div>
```
the generated output will just substitute `[object]` for that data.

It would be good if there were some documentation that could show how it's possible include / embed xml/svg data within a template.

**Comment by c-git** [comment]
Hopefully someone more experienced will provide a more conclusive answer but for me when I get `[object` I usually didn't access the property of the object that I wanted and needed to use dot notation to get one of the properties of the object. But sometimes I don't know what the properties are and therefore don't know how to start to try them to see which I want. And for that I use `{{ __tera_context }}` which shows all the data in the current context. Hope this helps in the mean while.

**Comment by rib** [comment]
thanks for the tip.

I think in this case it mainly just highlights the `load_data` is probably not the right tool for trying to include xml directly into a template. It looks like it essentially loads the xml data into an object whereby you can dig into the tags of the xml document (e.g. there is a data['svg'] element if I look at the `__tera_context` which I'm sure could be useful in some situations but in my case I'm more just wanting to read in the xml verbatim and include it at that location of the template.

**Comment by c-git** [comment]
Cool I've never used xml/svg like that. I don't know off hand how to do it but if you can create a sample repo I can clone it and give it a try. A quick Google on loading plain text came up with [this](https://zola.discourse.group/t/include-files-from-a-page/62) but not sure if it's of any help to you.

**Comment by c-git** [comment]
Have you already tried putting the svg into a shortcode?

**Comment by rib** [comment]
Not yet but maybe that'd be the way to go - I'll give it a try - I think originally that just conceptually seemed like the wrong tool but maybe not.

**Comment by rib** [comment]
Ah, no, shortcodes can't be used in templates (I'm trying to inline the svg icons as part of a footer that's in a base.html template)

**Comment by rib** [comment]
Okey, so I figured I could kind of use macros but it's really not very ergonomic unfortunately.

Firstly I can make a file named `svg-icons.xml` that looks something like:
```xml
{% macro github() %}
<svg fill="#000000" width="24px" height="24px" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
...
</svg>
{% endmacro github %}

{% macro linkedin() %}
<svg fill="#000000" width="24px" height="24px" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg">
...
</svg>
{% endmacro linkedin %}
```

Then I can add a line like this to the top of my template file:
`{% import "svg-icons.xml" as macros %}`

and then finally I can inline the SVG like:
```xml
<div class="icon">{{ macros::github() }}</div>
```

This feels kinda over the top though, considering the need to have to manually pack the content into a special macro file (which sort of defeats the aim)

**Comment by c-git** [comment]
Right, in the case I was dealing with tonight was in a markdown file so that's why I thought of shortcodes. Hope you find a solution or someone more experienced see it and knows a better way.

**Comment by rib** [comment]
thanks for the input + suggestions.

from what I can tell at the moment then macros may be the closest thing to what I was looking for but they aren't really a practical solution here. The general assumption is that there's a directory of pre-existing svg icons and it would be nice to be able to inline icons from their original `.svg` files, ideally without needing to pre-process them into a library of macros first.

**Comment by Keats** [comment]
The issue is that load_data loads it as structured data. To work with svg that way, we'd need to load data as plain text.

**Comment by rib** [comment]
Yeah, if there were maybe an option to load as text instead of as a structured object that could be good for this.

**Comment by rib** [comment]
I suppose I'd maybe just add that my first intuition was to try and use `{% include "foo.svg" %}` considering that I had previously used `{%- include "foo.svg" -%}` with liquid/jekyll.

Conceptually `load_data` seems like it's more geared for loading data that will be inserted into the generated page structure vs `include` being used to pull in content that itself forms the structure of the page.

**Comment by Keats** [comment]
`{% include "foo.svg" %}` works if you make sure you load svg files in your glob.

**Comment by rib** [comment]
(note that I was doing this from within a template - I thought I saw that `include` currently isn't supported in templates?)

**Comment by rib** [comment]
oh, no, wait, I'm getting confused with shortcodes there, ignore that

**Comment by rib** [comment]
I recall also trying to rename the files to be html/xml to see if there was an issue with the extension and couldn't get that to work, but maybe need to double check that.

**Comment by rib** [comment]
I don't currently see where I can set a glob pattern to affect Tera `include`s. Looking at the example config here https://www.getzola.org/documentation/getting-started/configuration/ I just see the pattern for `ignored_content` which isn't related to this.

I tried grepping the zola code quickly but didn't see something obvious.

It looks like glob patterns can be given to `Tera` for the template directories but I'm not sure atm how that's configurable from a zola site config.

**Comment by rib** [comment]
I can at least confirm that renaming the svg files to be `.xml` files does work as suggested - so when I originally tried this I guess I had the file in a non-template directory.

So yeah, if it's easy to extend the glob pattern to include `*.svg` that would be good - I can't currently see where I can do that?

**Comment by Keats** [comment]
Ah I was thinking as a Tera issue. We should probably include svg in the glob pattern as you say.

**Comment by rib** [comment]
okey, yeah that would be good for this yep

**Comment by timharek** [comment]
I've managed to do this for my personal site with `{{ load_data(path="path/to/svg.svg") | safe }}`.

You can check out the source code here: [timharek/timharek.no:/layouts/templates/base/header.html](https://github.com/timharek/timharek.no/blob/main/templates/base/header.html#L3)

**Comment by biodranik** [comment]
@Keats can you please also allow .js extensions in `include` statements? They don't work now. It is very convenient to include pure JS files from templates, and support js-related syntax highlighting and linters.

**Comment by Keats** [comment]
Maybe we can make the extensions configurable or just load everything in the templates directory? JS/SVG should not conflict with Tera syntax

**Comment by PSeitz** [comment]
> I've managed to do this for my personal site with `{{ load_data(path="path/to/svg.svg") | safe }}`.
> 
> You can check out the source code here: [timharek/timharek.no:/layouts/templates/base/header.html](https://github.com/timharek/timharek.no/blob/main/templates/base/header.html#L3)

Is there a way to forward parameters here? Currently I need a shortcode for each svg that is called in the `.md` file. 

```html
shortcode1.html
<div {% if class %}class="{{class}}"{% endif %}>
{{ load_data(path="static/file1.svg") | safe }}
</div>

shortcode2.html
<div {% if class %}class="{{class}}"{% endif %}>
{{ load_data(path="static/file2.svg") | safe }}
</div>
...
```

**Comment by dscso** [comment]
For my website I've written a little helper macro:
```jinja2
{%- macro icon(name, type) -%}

{% set data = load_data(path="themes/icons/svgs/"~type~"/"~name~".svg") %}
<svg class="icon">
{{ data | safe }}
</svg>

{%- endmacro -%}

{%- macro iconcss() -%}

<style>
.icon {
    width: 1em;
    height: 1em;
    vertical-align: -.142em;
    fill: currentColor;
}
</style>
{%- endmacro -%}
```

and I have downloaded the `svgs` folder from [FontAwesone](https://github.com/FortAwesome/Font-Awesome) into my little theme

Then I can use it like so:
```jinja2
<div>
    follow us on youtube {{ icon::icon(name = "github", type = "brands") }}
</div>
```

And I added the style macro somewere in the footer

```jinja2
    <!-- ... -->
    {{ icon::iconcss() }}
  </body>
</html>
```

Works well for me, but [Hugo icons](https://github.com/hugo-mods/icons) uses a cool approach, where it remembers which icons have been used. Then it includes these icons at the end of the HTML file so if you use the same icon multiple times, you just have the path once in the HTML file.

In HTML this looks something like this: When calling the `{{ icon "github" }}` it generates this:
```html
<svg class="icon"><use xlink:href="#icon-github"></use></svg>
```
then it adds at the bottom of the page this tag:
```html
<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
        <symbol id="icon-github" viewBox="0 0 496 512">
          <title>GitHub</title>
          
          <!-- Font Awesome Free 5.15.3 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"></path>
        </symbol>
</svg>
```
 

If there would be somehow a way to save state during each render, this could be easily implemented. Is someone else also interested in this?
